editor/scene_create_dialog.cpp

```diff
@@ -82,40 +81,35 @@ void SceneCreateDialog::browse_types() {
 void SceneCreateDialog::on_type_picked() {
 	other_type_display->set_text(select_node_dialog->get_selected_type().get_slice(" ", 0));
 	if (node_type_other->is_pressed()) {
-		update_dialog();
+		validation_panel->update();
 	} else {
-		node_type_other->set_pressed(true); // Calls update_dialog() via group.
+		node_type_other->set_pressed(true); // Calls validation_panel->update() via group.
 	}
 }
 
 void SceneCreateDialog::update_dialog() {
 	scene_name = scene_name_edit->get_text().strip_edges();
-	update_error(file_error_label, MSG_OK, TTR("Scene name is valid."));
 
-	bool is_valid = true;
 	if (scene_name.is_empty()) {
-		update_error(file_error_label, MSG_ERROR, TTR("Scene name is empty."));
-		is_valid = false;
+		validation_panel->set_message(0, TTR("Scene name is empty."), EditorValidationPanel::MSG_ERROR);
```

Comment:
So I have a question about the overall design of the new component.

It's expected that you, as a user, would update all messages in one go, within one call of the update callback, whatever triggers it. This makes me question whether it makes sense to rigidly set a predetermined list of possible status messages and create labels for them. Why not just clear everything and create the necessary lines, whether OK or WARN or ERROR on the fly, as we call the update callback? We can still evaluate if messages contain any errors and update the accept button, disable or enable it.

This seems a bit on the overengineered side of things.

----

That said, if you have a good argument for this approach, I would still suggest an adjustment. There is currently no clear connection between the messages that we update. We pass some collection to the constructor, and then here we set the message for some index 0. I would rework this so that instead of indices we use an enum/constants, so instead of some mysterious 0 we have a clear pointer to the message that we update, e.g. `SCENE_NAME_VALIDITY` or something.

And instead of passing an array to the constructor I would add a method for adding messages and use these constants there as well. So both when setting and updating it is clear which validation message we are touching. So it'd be like

```cpp

validation_panel->add_message(TTR("Some valid message"), SCENE_NAME_VALIDITY);
```

This would mean that instead of indices we use ids, but that's a fairly common pattern in our codebase.

Again, this is if you insist on doing it this way to begin with. But I think the whole scaffolding with labels is unnecessary and the update callback can set both good and bad message, or no message at all.

Reply:
> This makes me question whether it makes sense to rigidly set a predetermined list of possible status messages and create labels for them. Why not just clear everything and create the necessary lines, whether OK or WARN or ERROR on the fly, as we call the update callback?

If they are predetermined, the order is guaranteed, so it doesn't matter what message you set and when.
Also there is no reason to re-create and setup them every time, not to mention it can be costly (validation is called with every text change).

Reply:
The order is guaranteed anyway, since the update callback is the only source of state changes. So the order is determined by it as well (or was, before this PR).

I don't get what you mean by it being costly. We have the update callback; before calling it we set the state to false, during it we set it to true if an error is added, after it we do the necessary updates.

Reply:
> The order is guaranteed anyway, since the update callback is the only source of state changes.

Yeah, and then there is ScriptCreateDialog with its tangled update logic. Appending messages would be super error-prone.

I only added enum in one case, because there were many different messages; for simple dialogs it's not that important. But I could add it everywhere I guess; IDs make sense too. Main reason why I designed this system like this is that I wanted to minimize the code required to add validation to a dialog; lots of stuff happens automatically behind the scenes.

> I don't get what you mean by it being costly.

The update callback is called with LineEdit's `text_changed`, which means every time you type or remove a letter. It wouldn't have noticeable impact probably, but destroying and creating nodes this often is not very efficient.